driver-core/src/main/com/mongodb/operation/MixedBulkWriteOperation.java

```diff
@@ -203,64 +245,97 @@ public void call(final AsyncConnection connection, final Throwable t1) {
         });
     }
 
-    private BulkWriteResult executeBatches(final Connection connection, final SessionContext sessionContext) {
-        BulkWriteBatch batch = BulkWriteBatch.createBulkWriteBatch(namespace, connection.getDescription(), ordered,
-                writeConcern, bypassDocumentValidation, writeRequests);
+    private BulkWriteResult executeBulkWriteBatch(final WriteBinding binding, final Connection connection,
+                                                  final BulkWriteBatch originalBatch) {
+        BulkWriteBatch currentBatch = originalBatch;
+        MongoException exception = null;
 
-        while (batch.shouldProcessBatch()) {
-            BsonDocument result = connection.command(namespace.getDatabaseName(), batch.getCommand(), NO_OP_FIELD_NAME_VALIDATOR,
-                    ReadPreference.primary(), batch.getDecoder(), sessionContext, shouldAcknowledge(batch, writeConcern),
-                    batch.getPayload(), batch.getFieldNameValidator());
-            batch.addResult(result);
-            batch = batch.getNextBatch();
+        try {
+            while (currentBatch.shouldProcessBatch()) {
+                BsonDocument result = executeCommand(connection, currentBatch, binding);
+                currentBatch.addResult(result);
+                currentBatch = currentBatch.getNextBatch();
+            }
+        } catch (MongoException e) {
+            exception = e;
+        } finally {
+            connection.release();
         }
 
-        return batch.getResult();
+        if (exception == null) {
+            return currentBatch.getResult();
+        } else if (!currentBatch.shouldProcessBatch() || !originalBatch.getRetryWrites() || !isRetryableException(exception)) {
```

Comment:
I don't think shouldProcessBatch should ever be false here.  If so, remove from the conditional to avoid confusion.

Reply:
Done.